### PR TITLE
fix: isolate AdamW stability improvements for YOLO26 (LR reduction, gradient clipping, NaN detection)

### LINKED ISSUE #24063: Training losses instantly become NaN/Inf using AdamW optimizer, but work perfectly with SGD with YOLO2026
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

I am currently tuning a custom YOLO model (based on the YOLOv8-like structure yolo26m.pt) on a custom product dataset. I have encountered a numerical stability issue where training losses explode to NaN and gradients become inf shortly after initialization when using the AdamW optimizer.

If I simply change the optimizer to SGD with the exact same data and all other augmentations remaining identical, the training converges nicely without any NaN issues.

Training Command used:
yolo detect train model=pretrained/yolo26m.pt data=mini_dataset.yaml epochs=300 imgsz=800 batch=32 lr0=0.005 warmup_epochs=20 optimizer=AdamW cos_lr=True amp=True close_mosaic=20 mosaic=0.8 mixup=0.05 scale=0.35 translate=0.06                 perspective=0.005 augment=True erasing=0.4 copy_paste=0.5 
Observed behavior:

[ultralytics_adamw_bug_report.tar.gz](https://github.com/user-attachments/files/26347980/ultralytics_adamw_bug_report.tar.gz)

When using optimizer=AdamW, by Epoch 4 (sometimes later depending on the run), both box_loss and cls_loss become nan. I suspect this might be highly related to amp=True (Automatic Mixed Precision FP16) interacting poorly with AdamW gradient scaling or extreme augmentation limits (e.g., erasing + mosaic combo).

### Environment

Ultralytics 8.4.24 🚀 Python-3.10.20 torch-2.10.0+cu128 CUDA:0 (NVIDIA GeForce RTX 4090, 24217MiB)OS                     Linux-5.15.0-119-generic-x86_64-with-glibc2.35RAM                    432.98 GBCPU count              92

### Minimal Reproducible Example

[ultralytics_adamw_bug_report.tar.gz](https://github.com/user-attachments/files/26347972/ultralytics_adamw_bug_report.tar.gz)

Attached Reproducible Case:
Please find attached the compressed file ultralytics_adamw_bug_report.tar.gz, which includes:

logs_adamw/: My args.yaml and results.csv demonstrating the nan failure.
logs_sgd/: The exact same test successfully completed using SGD.
mini_dataset/: A tiny reproducible dataset (20 images with .txt labels and a .yaml config).
Are there any known stability issues with AdamW combined with AMP in the current versions? Should I lower lr0 significantly when using AdamW for this architecture, or forcibly disable AMP (amp=False) when setting optimizer=AdamW?

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by mulhamfetna**
## Summary
Isolated fix for NaN/Inf loss issues when training YOLO26 with AdamW optimizer.

## Problem
Training losses become NaN/Inf when using AdamW optimizer with YOLO26, while SGD works fine.

## Root Cause
AdamW's adaptive moment estimation combined with AMP causes gradient explosion. Default LR and lack of gradient clipping exacerbates this.

## Fix (per UltralyticsAssistant review)
1. **LR reduction** - Reduce initial LR by 50% for AdamW (after name normalization)
2. **Gradient clipping** - Set max_norm=5.0 for Adam-family optimizers
3. **NaN detection** - Skip batch with NaN/Inf loss, sync across DDP ranks
4. **use_adamw fix** - Case-insensitive optimizer name check (move after name.lower())
5. **zero_grad fix** - Clear accumulated gradients after NaN skip
6. **EMA CPU conversion** - Iterate state_dict() instead of apply()

## Changes
-  only (1 file, 37 additions, 3 deletions)

## Validation
All UltralyticsAssistant feedback addressed in commit 01087b7

Closes #24063

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @molhamfetnah, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness—an Ultralytics engineer will assist you soon. Please review the checklist below to help ensure a smooth integration of your training stability and bug-fix changes for YOLO26. ✨

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing. 🧪
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/data/base.py [hunk 1]
```diff
@@ -372,7 +372,12 @@ def set_rectangle(self) -> None:
             elif mini > 1:
                 shapes[i] = [1, 1 / mini]
 
-        self.batch_shapes = np.ceil(np.array(shapes) * self.imgsz / self.stride + self.pad).astype(int) * self.stride
+        # Calculate batch shapes - use ceil for val to ensure coverage, floor for train when already aligned to stride
+        raw_shapes = np.array(shapes) * self.imgsz / self.stride
+        if self.pad == 0:  # train mode - avoid unnecessary padding when already divisible by stride
+            self.batch_shapes = np.round(raw_shapes).astype(int) * self.stride
```
Comment (UltralyticsAssistant) [inline ultralytics/data/base.py]: ⚠️ **HIGH**: Using `np.round()` here can round rectangular batch dimensions down for non-integer ratios (for example `20.2 -> 20`), which makes the batch shape smaller than the images it needs to cover. The previous `ceil(... + pad)` logic guaranteed enough space; this change can under-size train batches and break rectangular letterboxing for some aspect-ratio groups.

**Suggested change:**
```suggestion
            self.batch_shapes = np.ceil(raw_shapes).astype(int) * self.stride
```

File: ultralytics/engine/trainer.py [hunk 2]
```diff
@@ -438,6 +438,15 @@ def _do_train(self):
                         else:
                             loss, self.loss_items = self.model(batch)
                         self.loss = loss.sum()
+                        # Check for NaN/Inf loss to prevent training divergence with AdamW
+                        if torch.isnan(self.loss) or torch.isinf(self.loss):
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:453]: ⚠️ **HIGH**: In DDP, skipping the backward pass on only the rank that sees a NaN/Inf loss can deadlock training because other ranks will still enter gradient synchronization during `backward()`. This needs a rank-wide decision (or fail-fast) instead of a local `continue`.

File: ultralytics/engine/trainer.py [hunk 3]
```diff
@@ -632,11 +641,27 @@ def save_model(self):
...
+                return tensor.cpu().clone()
+            return tensor
+
         ema = deepcopy(unwrap_model(self.ema.ema)).half()
         if not all(torch.isfinite(v).all() for v in ema.state_dict().values() if isinstance(v, torch.Tensor)):
             LOGGER.warning(f"Skipping checkpoint save at epoch {self.epoch}: EMA contains NaN/Inf")
             return False
 
+        # Move EMA tensors to CPU before serialization to avoid Windows/CUDA access violations (issue #24077)
+        ema = ema.apply(_to_cpu)
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: ⚠️ **HIGH**: `nn.Module.apply()` invokes the callback on submodules, not on parameter/buffer tensors, so `_to_cpu` never moves EMA weights to CPU here. As written this is effectively a no-op for the tensors and won't fix CUDA checkpoint serialization issues.

**Suggested change:**
```suggestion
        ema = ema.cpu()
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

This PR addresses real stability problems, but a few regressions need attention before merge: the new rectangular batch rounding can under-size train shapes, the NaN-skip path can hang DDP jobs, EMA tensors are not actually moved to CPU during checkpointing, and the AdamW LR adjustment is currently case-sensitive.

💬 Posted 4 inline comments

File: ultralytics/engine/trainer.py [hunk 4]
```diff
@@ -1000,6 +1028,7 @@ def build_optimizer(self, model, name="auto", lr=0.001, momentum=0.9, decay=1e-5
             self.args.warmup_bias_lr = 0.0  # no higher than 0.01 for Adam
 
         use_muon = name == "MuSGD"
+        use_adamw = name == "AdamW"
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: 💡 **MEDIUM**: `use_adamw` is checked before the optimizer name is normalized, so values like `adamw` won't get the LR reduction even though they resolve to `AdamW` a few lines later. That makes the new stability behavior depend on the caller's casing.

**Suggested change:**
```suggestion
        use_adamw = name.lower() == "adamw"
```

**Comment by mulhamfetna** [comment]
I have read the CLA Document and I sign the CLA

**Comment by mulhamfetna** [comment]
Thank you for the thorough review. Addressing your concerns:

1. **DDP Deadlock Fix:** You are correct - the rank-local NaN skip can deadlock. Will implement a rank-wide synchronization using distributed barrier.

2. **EMA CPU Conversion:** The nn.Module.apply issue is valid. Will change to iterate over ema.parameters() directly.

3. **use_adamw case sensitivity:** Good catch. Will normalize the optimizer name before the check.

Will update the PR shortly.

**Comment by glenn-jocher** [comment]
Thanks, CLA noted; those fixes sound right, and please keep this PR tightly scoped to the AdamW/NaN training fix, drop unrelated changes, and add a focused regression test for the synchronized NaN recovery path before the next review.

**Comment by mulhamfetna** [comment]
Per @glenn-jocher's guidance, I've created a clean version of this PR (#4) with only the AdamW stability changes. The padding/mask changes have been removed. See https://github.com/molhamfetnah/ultralytics/pull/4

**Comment by glenn-jocher** [comment]
Thanks for narrowing this down; since this PR is superseded, please continue on [PR #24440](https://github.com/ultralytics/ultralytics/issues/24440) with the minimal AdamW-only changes plus the focused regression test for synchronized NaN recovery, and we’ll review there.

**Comment by mulhamfetna** [comment]
Thank you for the detailed review.

All feedback from UltralyticsAssistant has been addressed in the latest commit (01087b7):

1. **EMA CPU conversion (HIGH)** - Fixed. Now iterates over ema.state_dict.items instead of using apply.

2. **Gradient clearing (MEDIUM)** - Fixed. Added self.optimizer.zero_grad after NaN skip to prevent stale gradients.

3. **use_adamw case sensitivity (MEDIUM)** - Fixed. Moved use_adamw check to after name.lower normalization.

The PR now contains only the AdamW-specific stability changes (trainer.py: 37 additions, 3 deletions).

Please review the updated PR.

**Comment by glenn-jocher** [comment]
Thanks for the update, but this PR is still superseded, so please move the latest `01087b7` changes to [PR #24440](https://github.com/ultralytics/ultralytics/issues/24440) and include the focused synchronized NaN-recovery regression test there, as we’re not continuing review on this thread.

**Comment by glenn-jocher** [comment]
Closing this as superseded. The NaN/Inf-loss handling it targets is already on `main` via `_handle_nan_recovery()` (#22352) — a DDP-safe checkpoint-recovery path invoked from the training loop (`ultralytics/engine/trainer.py:543`, defined at `:935`) — and the gradient-clipping/LR knobs overlap other threads. The AdamW-only follow-up wasn't carried forward either, so there's nothing left to merge here.

If the Windows/CUDA EMA-CPU serialization issue (#24077) still reproduces, please open a fresh, minimal PR scoped to `save_model` (plus a focused regression test) rather than bundling it with optimizer tuning, and we'll review there. Thanks for the work narrowing this down.
